src/components/toolbar/TypeSelector.tsx

```diff
@@ -0,0 +1,437 @@
+          previousNode,
+          updatedDeepestSubsection
+        )
+        fragment = updatedPreviousNode.content
+      } else {
+        // If no subsections exist, just append to the main section
+        fragment = Fragment.from(previousNode.content)
+          .append(Fragment.from(paragraph))
+          .append(sectionContent)
+      }
+
+      tr.replaceWith(
+        beforeSection - previousNode.nodeSize,
+        afterSection,
+        previousNode.copy(fragment)
+      )
+
+      anchor = beforeSection
+    } else {
+      tr.replaceWith(
+        beforeSection,
+        afterSection,
+        Fragment.from(paragraph).append(sectionContent)
+      )
+
+      anchor = beforeSection + 1
+    }
+
+    tr.setSelection(
+      TextSelection.create(tr.doc, anchor, anchor + paragraph.content.size)
+    )
+
+    dispatch(skipTracking(tr))
+    view && view.focus()
+  }
+
+  const parentElementType = parentElement.node.type
+  const options: Option[] = []
+
+  switch (parentElementType) {
+    case parentElementType.schema.nodes.section: {
+      const sectionDepth = Math.max(1, $from.depth - 1)
+      const beforeSection = $from.before(sectionDepth)
+      const $beforeSection = doc.resolve(beforeSection)
+      const sectionOffset = $beforeSection.parentOffset
+
+      options.push(
+        {
+          nodeType: nodes.paragraph,
+          label: 'Paragraph',
+          action: demoteSectionToParagraph,
+          isDisabled: sectionDepth <= 1 && sectionOffset <= 1,
+          isSelected: false,
+        },
+        {
+          nodeType: nodes.section,
+          label: 'Heading',
+          isDisabled: true,
+          isSelected: true,
+        }
+      )
+      return options
+    }
+    case parentElementType.schema.nodes.paragraph: {
+      options.push(
+        {
+          nodeType: nodes.paragraph,
+          label: 'Paragraph',
+          isDisabled: true,
+          isSelected: true,
+        },
+        {
+          nodeType: nodes.section,
+          label: 'Heading',
+          action: promoteParagraphToSection(),
+          isDisabled: false,
+          isSelected: false,
+        }
+      )
+      return options
+    }
+    default: {
+      options.push({
+        nodeType: parentElementType,
+        label: titleCase(optionName(parentElementType)),
+        isDisabled: true,
+        isSelected: true,
+      })
+      return options
+    }
+  }
+}
+
+const StyledSelect = styled(Select<Option, false>)`
+  & > div:hover {
+    border-color: ${(props) => props.theme.colors.border.secondary};
+  }
+`
+
+const findSelectedOption = (options: Option[]): Option | undefined => {
+  for (const option of options) {
+    if (option.isSelected) {
+      return option
+    }
+  }
+}
+
+export const TypeSelector: React.FC<{
+  state: ManuscriptEditorState
+  dispatch: (tr: Transaction) => void
+  view?: ManuscriptEditorView
+}> = ({ state, dispatch, view }) => {
+  const options = buildOptions(state, dispatch, view)
```

Comment:
Do you think we can benefit from memoising the buildOptions here? The buildOptions function is relatively big and  if it will experience rerendering more often than _state_ or _view_ change, we should memoise it.

Reply:
I can't see a real benefit from memorizing it, specially because `TypeSelector` will not re-render frequently without changes in `state`.